### PR TITLE
feat: Add yay provider

### PR DESCRIPTION
**Pull request by knelasevero**
Hey! Very cool project! :smile:

Was setting up an old laptop at home and came across this very promising project. My wife and I are using Arch/Manjaro and having support to yay/pacman would be great!

### CODE REVIEW

**Comment by rawkode** [comment]
Hey. Thanks for the PR. I'll get that code coverage action disabled on forks, that's never going to work because of the secret. 

The bootstrap function uses pacman to install Yay, but it's only available in the AUR (maybe different on Manjaro?). I'm actually OK with the bootstrap returning an error for now and leaving a message that says to manually install yay, as it's not easy to coordinate multiple steps with command execution yet, or raise privilege; though this will be ready by tomorrow #50

Thanks again 😀

**Comment by rawkode** [comment]
Closes #16

**Comment by knelasevero** [comment]
You are right! Forgot that in a pure arch box  you have some extra steps to have an AUR helper available :sweat_smile: 

I added more commands if you want to merge this one before the command privilege escalation PR. I can also wait for the other one to get merged and rebase it here if you prefer :)

Tested this in a pure Arch machine and it seems to work.

**Comment by knelasevero** [comment]
Saw that you merged it just now, already rebased it here and started using the new command approach

**Comment by rawkode** [comment]
Damn, that was fast! The new `run_command` helper should be much easier for supporting new providers like this. I'll get this merged when you're happy with it 😀

File: src/actions/package/providers/yay.rs
Comment (rawkode) [inline src/actions/package/providers/yay.rs:89]: `yay` should be run as regular user I believe, but it does require privilege elevation for the installation. 

Hmm, so I guess I may need to make the `run_command` helper provide a way to inherit IO without sudo?

Or am I mistaken?

**Review by rawkode — CHANGES_REQUESTED** [review]

File: src/actions/package/providers/yay.rs
Comment (rawkode) [reply src/actions/package/providers/yay.rs:90]: In fact, this can just be false. The require root for the commands above will mean we still have an active sudo session and won't need a password.

Comment (knelasevero) [reply src/actions/package/providers/yay.rs]: Makes sense :)

File: src/utils/command.rs
Comment (knelasevero) [inline src/utils/command.rs:46]: I am a Rust newbie, what would be a good way to set this current_dir only if dir is set?

Something like setting it to current dir if it is not set and setting it to command.dir if it is set?

Comment (rawkode) [reply src/utils/command.rs:46]: You could make the `dir` field on the struct `Option<String>` and then use `.current_dir(&command.dir.clone().unwrap_or(std::env::current_dir())`

Comment (rawkode) [reply src/utils/command.rs:46]: The more idiomatic way would be to `impl Default` for the `Command` struct and provide defaults. It's a little more work.

https://doc.rust-lang.org/std/default/trait.Default.html

Comment (rawkode) [reply src/utils/command.rs:46]: I'm happy with either approach, things can be cleaned in due course. Lets get you this feature shipped 😄

File: src/actions/command/run.rs
Comment (rawkode) [inline src/actions/command/run.rs:34]: `dir: Some(self.dir.clone())` now that it's an option value

File: src/actions/package/providers/mod.rs
Comment (rawkode) [inline src/actions/package/providers/mod.rs:21]: Shall we add pacman too?

`#[serde(alias = "yay", alias = "pacman")]`

File: src/actions/package/providers/yay.rs
Comment (rawkode) [inline src/actions/package/providers/yay.rs:5]: This conflicts with `Command` above. I don't believe we need this import anymore. It can be removed.

File: src/actions/package/providers/yay.rs
Comment (rawkode) [inline src/actions/package/providers/yay.rs:33]: `self.env()` isn't defined on this  `impl`.

You can use `HashMap::new()` instead, to keep it an empty environment

File: src/actions/package/providers/yay.rs
Comment (rawkode) [inline src/actions/package/providers/yay.rs:46]: Same as above

File: src/actions/package/providers/yay.rs
Comment (rawkode) [inline src/actions/package/providers/yay.rs:58]: Same as above

File: src/actions/package/providers/yay.rs
Comment (rawkode) [inline src/actions/package/providers/yay.rs:60]: `dir: Some(String::from("/tmp/yay"))`

File: src/actions/package/providers/yay.rs
Comment (rawkode) [inline src/actions/package/providers/yay.rs:84]: Same as above

**Review by rawkode — CHANGES_REQUESTED** [review]

**Comment by rawkode** [comment]
Sorry, I didn't realise you were new to Rust. I've gone through and tried to identify all the little pesky mistakes.

Let me know if you want me to push any fixes instead 👍🏻

**Comment by knelasevero** [comment]
> Sorry, I didn't realise you were new to Rust. I've gone through and tried to identify all the little pesky mistakes.
> 
> Let me know if you want me to push any fixes instead 👍🏻

Thanks! Yeah, only touched Rust before in Advent of code challenges :smile: Gonna eat something and fix all the details!

**Comment by knelasevero** [comment]
Also sorry for Pushing code with compilation errors. I am pulling this in a fresh vm to test in a place without the AUR helper 😅 gonna tidy up in a bit

**Comment by rawkode** [comment]
> Also sorry for Pushing code with compilation errors. I am pulling this in a fresh vm to test in a place without the AUR helper 😅 gonna tidy up in a bit

Not a problem, I'm here if you need any help. Have fun!

File: src/utils/command.rs
Comment (knelasevero) [inline src/utils/command.rs:51]: I think that the only thing missing now is this. I had to `.unwrap().into_os_string().into_string().unwrap()` because `current_dir()` gives us a PathBuf. But I think I am screwing something up here in the process. It seems that command is always executing in `std::env::current_dir()` anyways

Comment (rawkode) [reply src/utils/command.rs:51]: I'll take a look 👍

File: src/actions/package/providers/yay.rs
Comment (rawkode) [inline src/actions/package/providers/yay.rs:88]: Use `None` when we don't mind where this gets executed.

**Review by rawkode — CHANGES_REQUESTED** [review]

File: src/utils/command.rs
Comment (rawkode) [reply src/utils/command.rs:51]: It looks fine. What output do you see when you run it? Sorry, I don't have an Arch machine atm

Comment (knelasevero) [reply src/utils/command.rs:51]: Yeah, this was not the problem. It seems that if we try to use `require_root: true` for users that have `NOPASSWD: ALL` nothing breaks but commands are not executed. Same thing for running on shell that recently executed sudo stuff.

Comment (knelasevero) [reply src/utils/command.rs]: Tried this on a machine that I have with ubuntu to check the apt provider, and even though comtrya runs fine, in the end packages are not installed. Not sure why `require_root: true` is making it silently fail.

Comment (rawkode) [reply src/utils/command.rs]: Hmm. I'll dig in to that. Thanks for letting me know

Comment (rawkode) [reply src/utils/command.rs]: I'm going to add some end to end tests to the CI job to ensure this doesn't happen again, and I'll add assertion support to manifest 👍

**Comment by rawkode** [comment]
Are you happy with this now?

File: src/utils/command.rs
Comment (knelasevero) [reply src/utils/command.rs]: Awesome!

**Comment by knelasevero** [comment]
> Are you happy with this now?

I am happy :) I think that after CI we can merge. Thanks for reviewing @rawkode!

**Comment by rawkode** [comment]
Perfect. I'll dig in to the sudo thing this evening. Thanks for the awesome PR

**Review by rawkode — APPROVED** [review]

**Comment by rawkode** [comment]
Stupid bug on my part, which I believe caused the errors you mentioned. Sorry about that!

dabfb77

@knelasevero

**Comment by knelasevero** [comment]
😹 no worries! Cool that you found it!

**Comment by knelasevero** [comment]
Btw, @rawkode , if you need anything tested on arch I can happily do that here ;)

**Pull request merged**
